BacktraceScreen.c

```diff
@@ -0,0 +1,419 @@
+   if (!basename || !object) {
+      return;
+   }
+
+   char* lastSlash = strrchr(object, '/');
+   if (lastSlash) {
+      object = lastSlash + 1;
+   }
+
+   char* objectTrimmed = String_trim(object);
+   if (String_eq(basename, objectTrimmed)) {
+      RichString_setAttrn(out, CRT_colors[PROCESS_BASENAME], object - line, strlen(objectTrimmed));
+   }
+
+   free(objectTrimmed);
+}
+
+static void BacktraceFrame_display(const Object* super, RichString* out) {
+   const BacktraceFrame* frame = (const BacktraceFrame* )super;
+   assert(frame);
+
+   if (frame->index == BACKTRACE_FRAME_HEADER_INDEX) {
+      BacktraceFrame_displayHeader(frame, out);
+      return;
+   }
+
+   const BacktracePanelPrintingHelper* printingHelper = &frame->backtracePanel->printingHelper;
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   char* functionName = frame->functionName;
+   int maxFunctionNameLength = printingHelper->maxFuncNameLen;
+   if (!!(displayOptions & DEMANGLE_NAME_FUNCTION) &&
+         printingHelper->maxDemangledFuncNameLen > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFuncNameLen;
+      if (frame->demangleFunctionName) {
+         functionName = frame->demangleFunctionName;
+      }
+   }
+
+   char *completeFunctionName = NULL;
+   xAsprintf(&completeFunctionName, "%s+0x%zx", functionName, frame->offset);
+
+   char* objectDisplayed = frame->objectName;
+   size_t objectLength = printingHelper->maxObjNameLen;
+   if (!!(displayOptions & SHOW_FULL_PATH_OBJECT)) {
+      objectDisplayed = frame->objectPath;
+      objectLength = printingHelper->maxObjPathLen;
+   }
+
+   size_t maxAddrLen = printingHelper->maxAddrLen - strlen("0x");
+   char* line = NULL;
+   int objectPathStart = -1;
+
+   /*
+    * The parameters for printf are of type int.
+    * A check is needed to prevent integer overflow.
+    */
+   assert(printingHelper->maxFrameNumLen <= INT_MAX);
+   assert(maxAddrLen <= INT_MAX);
+   assert(maxFunctionNameLength <= INT_MAX);
+   assert(objectLength <= INT_MAX);
+   int len = xAsprintf(&line, "%*d 0x%0*zx %-*s %n%-*s",
+      (int)printingHelper->maxFrameNumLen, frame->index,
+      (int)maxAddrLen, frame->address,
+      (int)maxFunctionNameLength, completeFunctionName,
+      &objectPathStart,
+      (int)objectLength, objectDisplayed ? objectDisplayed : "-"
+   );
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   free(completeFunctionName);
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* frame1 = (const BacktraceFrame*)object1;
+   const BacktraceFrame* frame2 = (const BacktraceFrame*)object2;
+   return String_eq(frame1->functionName, frame2->functionName);
```

Comment:
```suggestion
   return SPACESHIP_NULLSTR(frame1->functionName, frame2->functionName);
```

Reply:
Yes, good catch. I just learned, that when a `NULL` pointer is passed to `strcmp`, it's an undefined behavior.